File: lidar/lidar.py
import time
import mraa

millis = lambda: int(round(time.time() * 1000))


class Lidar(object):
    '''class to represent vx-11 lidar'''

    distances = [0] * 360
    packet = [0] * 22

    startReading = False
    packetNumber = 0
    speed = 0
    packetIndex = 0
    distanceIndex = 0

    def __init__(self, port=0, baudrate=115200):
        self.port = port
        self.uart = mraa.Uart(self.port)
        self.uart.setBaudRate(baudrate)
        self.count = 0

    def read(self):
        if self.uart.dataAvailable():
            byte = self.uart.read(1)
            byte = int(ord(byte))

            if self.startReading:
                if self.packetIndex == 21:
                    self.startReading = False
                    self.processEndOfPacket()
                    # let the calling function know we're done
                    if self.distanceIndex >= 359:
                      self.count += 1
                      return True
                    return False
                else:
                    self.packet[self.packetIndex] = byte
                    self.packetIndex += 1
                    return False

            if not self.startReading and byte == 0xFA:
                self.startReading = True
                self.packetIndex = 0
                self.packet[self.packetIndex] = byte
                self.packetIndex += 1
                return False

    def processEndOfPacket(self):
        newPacketNumber = self.packet[1]

        if (newPacketNumber - self.packetNumber) == 1:
            for i in range(22):
                if i == 2:
                    self.speed = self.packet[3] << 8 | self.packet[2]
                    self.speed /= 64
                elif self.isDataIndex(i):
                    self.distanceIndex = (newPacketNumber - 0xA0) * 4 + i/4 - 1
                    if self.distanceIndex >= 0 and self.distanceIndex < 360:
                        if (self.packet[i+1] & 0x80) >> 7:
                            self.distances[self.distanceIndex] = 0
                        else:
                            d = self.packet[i] | ((self.packet[i+1] & 0x3F) << 8)

                            # cap between limits
                            if d < 0 or d > 6000:
                                d = 0

                            # finally, store valid data
                            self.distances[self.distanceIndex] = d

                else:
                    #checksum, packet num, and start bit go here
                    pass


        self.packetNumber = newPacketNumber
        if self.packetNumber >= 0xF9:
            self.packetNumber = 0x9F

    def isDataIndex(self, index):
        return index == 4 or index == 8 or index == 12 or index == 16

    def getLatestDistances(self):
        return self.distances

File: lidar/lidar.cpp
#include "lidar.hpp"
#include "stdio.h"
#include "iostream"

Lidar::Lidar() try : dev(0),
    packetNumber(0),
    packetIndex(0),
    distanceIndex(0),
    startReading(false) {

  if (dev.setBaudRate(115200) != mraa::SUCCESS) {
       std::cout << "Error setting parity on UART" << std::endl;
   }
   if (dev.setMode(8, mraa::UART_PARITY_NONE, 1) != mraa::SUCCESS) {
       std::cout << "Error setting parity on UART" << std::endl;
   }
   if (dev.setFlowcontrol(false, false) != mraa::SUCCESS) {
       std::cout << "Error setting flow control UART" << std::endl;
   }
}
  catch (std::exception) {
    std::cout << "Error initialize uart" << std::endl;
    std::terminate();
}

bool Lidar::read(){
  if (dev.dataAvailable()) {
    char b;
    int read = dev.read(&b, 1);

    if (startReading) {
      if (packetIndex == 21) {
        startReading = false;
        processEndOfPacket();

        return distanceIndex >= 359;
      }
      else {
        packet[packetIndex] = b;
        packetIndex++;
      }
    }

    if (!startReading && (b == (char) 0xfa)){
      startReading = true;
      packetIndex = 0;
      packet[packetIndex] = b;
      packetIndex++;
      return false;
    }
  }
}

void Lidar::processEndOfPacket(){
  int newPacketNumber = packet[1];

  if ((newPacketNumber - packetNumber) == 1){
    for (int i=0; i < 22; i++){
      if (isDataIndex(i)){
        distanceIndex = (newPacketNumber - (char)0xA0) * 4 + i/4 - 1;
        if ((distanceIndex >= 0) && (distanceIndex < 360)){
          if ((packet[i+1] & (char)0x80) >> 7){
            distances[distanceIndex] = 0;
          }
          else {
            int d = packet[i] | ((packet[i+1] & (char)0x3f) << 8);

            if (d < 0 || d > 6000){
              d = 0;
            }

            distances[distanceIndex] = d;
            printf("%d, %d\n", distanceIndex, d);
          }
        }
      }
    }
  }

  packetNumber = newPacketNumber;

  if (packetNumber >= (char)0xF9){
    packetNumber = (char)0x9F;
  }
}

bool Lidar::isDataIndex(int index){
  return index == 4 || index == 8 || index == 12 || index == 16;
}

int main(int argc, char **argv){
  Lidar lidar;
  while (true) {
    lidar.read();
  }
}
